File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
      std::unordered_map<TreeNode*, std::pair<int, int>> map;
      map.insert({nullptr, {0, 0}});
      std::stack<TreeNode*> stack;

      stack.push(root);

      while (!stack.empty()) {
        auto current = stack.top();

        if (current->left != nullptr && map.find(current->left) == map.end()) {
          stack.push(current->left);
        } else if (current->right != nullptr && map.find(current->right) == map.end()) {
          stack.push(current->right);
        } else {
          current = stack.top();
          stack.pop();
          auto leftPair = map[current->left];
          auto rightPair = map[current->right];
          int height = 1 + std::max(leftPair.first, rightPair.first);
          int diameter = std::max(
            leftPair.first + rightPair.first,
            std::max(
              leftPair.second, rightPair.second
            )
          );

          map.insert({current, {height, diameter}});
        }
      }

      return map[root].second;
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.java
class Solution {
  public int maxProfit(int[] prices) {
    int left = 0;
    int right = 0;
    int maxDiff = 0;

    while (right != prices.length) {
      maxDiff = Math.max(maxDiff, prices[right] - prices[left]);
      
      if (prices[right] < prices[left]) 
        left = right;

      else 
        right++;
    }

    return maxDiff;
  }
}
